Add blog post on writing agent skills for datafusion-python (#187)
apache/datafusion-site: content/blog/2026-05-22-writing-agent-skills.md

---
layout: post
title: Writing Agent Skills for an Open Source Project: Lessons from DataFusion Python
date: 2026-05-22
author: Tim Saucer (rerun.io)
categories: [tutorial]
---

<!--
{% comment %}
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

[TOC]

> If you maintain an open source project, a growing fraction of the people

Reviewer: ```suggestion
If you maintain an open source project, a growing fraction of people
```

> using your library are not typing code directly anymore — they are asking an

Reviewer: ```suggestion
using your library are not typing code anymore — they are asking an
```

agent to write it for them. That agent leans on whatever it picked up during
training, which is rarely the idiomatic style your project actually wants.
The result is code that runs but reads like a stranger wrote it, or code
that doesn't run at all because the agent guessed at an API that doesn't
exist. You can fix this from inside the repository, with a small number of
**agent skills** checked in alongside your code.

This post is about how we did exactly that in
[`datafusion-python`][repo]. The specifics — DataFrame APIs,
PyO3-wrapped Rust bindings, an analytics library written on top of Apache
Arrow — are particular to our project, but the techniques are not. The
question of *who* a skill is for and how that shapes its contents, the
question of *where* a skill should live in the repo so the right people
load it, the question of *how to keep it honest* as your API evolves, and
the question of *how to evaluate it* against a real workload all generalize
to almost any library complex enough that an agent will struggle with it.

Concretely, you will get out of this post:

- A pattern for splitting skills by audience — user-facing vs.

>   contributor-facing — and why the split matters more than it sounds.

Reviewer: ```suggestion
  contributor-facing — and why the split matters.
```

- A workflow for keeping skills in sync with a moving API by treating the

>   skill itself as the maintenance tool.

Reviewer: ```suggestion
  skill itself as a maintenance tool.
```

- A method for grounding the user-facing skill against a corpus of known
  problems with known answers, run in a way that actually tests the skill
  instead of the agent's memory.
- A set of habits for evaluating and iterating on skills that apply to any
  project doing this work.

[repo]: https://datafusion.apache.org/python/

## What is an Agent Skill?

---

> A skill is a Markdown file (conventionally `SKILL.md`) with YAML frontmatter

Reviewer: My understanding is that agent skills is the current somewhat agreed upon standard for the skill structure, so the reference is nice IMO.
```suggestion
A [skill] is a Markdown file (conventionally `SKILL.md`) with YAML frontmatter
```

---

The single most important decision we made was to split skills into **two
clearly separate audiences**.

**End users** of `datafusion-python` are people writing application code:
loading Parquet files, building DataFrame queries, computing aggregates,
calling window functions. They want the agent to produce idiomatic
`SessionContext` / `DataFrame` / `Expr` code that runs on their data.

**Developers** of `datafusion-python` are the maintainers of the library
itself: people adding bindings, syncing with upstream [Apache DataFusion],
auditing API coverage, and refining the Python ergonomics of the
PyO3-wrapped Rust code. They want the agent to help them find gaps in the
binding layer and apply the fixes.

These two audiences need almost disjoint guidance. A user does not need to
know that `python/datafusion/functions.py` wraps `crates/core/src/functions.rs`,
or how to grep `~/.cargo/registry` for the upstream `invoke_with_args()`
implementation. A maintainer does not need a SQL-to-DataFrame migration
table. Mixing the two produces a skill that is too long for both audiences
and unfocused for either.

The other reason to keep them separate is **load semantics**. Skills are

> loaded into the model's context window. Every kilobyte of skill consumes

Reviewer: The kilobytes -> tokens mapping felt weird
```suggestion
loaded into the model's context window. Unnecessary skill detail consumes
```

  reach for `col(...)` only when the projection needs an expression.
- `HAVING` is the `filter=` keyword on the aggregate function, not a
  post-aggregation `filter()` call.
- Semi/anti joins instead of `EXISTS` / `NOT EXISTS` correlated
  subqueries.

These idiomatic rules are not in the user guide as a flat list — they
are scattered across docstrings, examples, and the implicit knowledge of
the maintainers. They show up in the skill because we watched an agent
write the non-idiomatic version and then went and wrote the rule down.
The contents of this list are not a property of `datafusion-python`;
they are a property of *what agents guess when they haven't seen your
library before*, and the only way to discover it is to put the skill in
front of a fresh agent and watch.

A useful trick during pass 3: when the agent does get something right
in a non-obvious way, ask it *why*. If the answer references something
that is not in your draft skill — a docstring it found, a public docs
page, a pattern from a similar library — that is a hint that the skill
is silent on something it should cover. Codify the reasoning, don't

> rely on the agent finding it again next time.

Reviewer: IMO this also helps to surface inconsistency in the source code base etc to resolve which is a valuable part of the skill development. If the agent found and copied a pattern you don't like it also motivates potentially resolve the source of that.

> The next two sections describe two different things we did after the

Reviewer: Two twos felt redundant
```suggestion
The next two sections describe different things we did after the
```

initial draft: a one-time grounding exercise against the TPC-H corpus
to validate the skill end-to-end, and a set of developer-side skills
that flag user-skill drift whenever the API moves.

## Grounding the Skill: TPC-H as a One-Time Validation

---

A draft skill needs to be tested against something more demanding than the
ad-hoc prompts the author used while writing it. We needed a way to confirm

> that the skill, once handed to a fresh agent, actually produced code that

Reviewer: ```suggestion
that the skill, once handed to a fresh agent, actually produces code that
```

Reviewer: There are mixed tenses between past work then current validation. I think this aligns them consistently

> *ran* and returned *correct answers* on real workloads — not just on the

Reviewer: ```suggestion
*runs* and returns *correct answers* on real workloads — not just on the
```

through each of the 22 generated scripts by hand and worked with the agent
to refactor them into the style the skill is supposed to teach: plain
column names where possible, `filter=` on aggregates instead of
post-aggregation filters, semi/anti joins instead of `EXISTS`, and so on.

Every time we caught the agent reaching for a non-idiomatic pattern, we
asked the same question: *did the skill teach this, or did the agent
infer it?* When the answer was "inferred," that was a gap in the skill, and
we updated `SKILL.md` to close it.

## The Developer Skills

---

The user skill exists to teach agents how to write good user code. The
developer skills, in `.ai/skills/`, exist to help maintainers keep the
project itself in good shape.

We ended up with three of them. The number was not planned up front;
each skill was written in response to a recurring chore that a
maintainer kept doing by hand and getting wrong in the same ways every
time. Once a task has a predictable shape and a checklist that a careful
person would follow, it is a candidate for a skill — and the act of
writing the skill forces you to make the checklist explicit.

> The three correspond to the three places maintenance drift shows up in a

Reviewer: ```suggestion
The skills correspond to the three places maintenance drift shows up in a
```

binding project like ours:

> - **`check-upstream`** — *the public API of the wrapped library moved

Reviewer: wrapped library felt ambiguous if we were talking about the library that does the wrapping or the source
```suggestion
- **`check-upstream`** — *the public API of the source library moved
```
